File: Sliding Window Algorithms/LeetCode 0076.py
from collections import Counter, defaultdict

class Solution:
    def minWindow(self, s: str, t: str) -> str:
        m, n = len(s), len(t)
        if n > m:
            return ""

        mp2 = Counter(t)
        mp1 = defaultdict(int)

        left = 0
        right = 0
        minlen = float('inf')
        start = 0
        check = 0

        while right < m:
            ch = s[right]
            mp1[ch] += 1
            if ch in mp2 and mp1[ch] == mp2[ch]:
                check += 1

            while check == len(mp2):
                if right - left + 1 < minlen:
                    minlen = right - left + 1
                    start = left
                mp1[s[left]] -= 1
                if s[left] in mp2 and mp1[s[left]] < mp2[s[left]]:
                    check -= 1
                left += 1

            right += 1

        return "" if minlen == float('inf') else s[start:start + minlen]

File: Sliding Window Algorithms/LeetCode 0076.java
import java.util.*;

class Solution {
    public String minWindow(String s, String t) {
        int m = s.length();
        int n = t.length();
        if(n > m)
        {
            return "";
        }
        HashMap<Character, Integer> mp1 = new HashMap<>();
        HashMap<Character, Integer> mp2 = new HashMap<>();

        for(char c : t.toCharArray())
        {
            mp2.put(c, mp2.getOrDefault(c, 0) + 1);
        }
        int left = 0;
        int right = 0;
        int start = 0;
        int minLen = Integer.MAX_VALUE;
        int check = 0;

        while(right < m)
        {
            char ch = s.charAt(right);
            mp1.put(ch, mp1.getOrDefault(ch, 0) + 1);
            if(mp2.containsKey(ch) && mp1.get(ch).intValue() == mp2.get(ch).intValue())
            {
                check++;
            }
            while(check == mp2.size())
            {
                if(right - left + 1 < minLen)
                {
                    minLen = right - left + 1;
                    start = left;
                }
                char leftChar = s.charAt(left);
                mp1.put(leftChar, mp1.get(leftChar) - 1);
                if(mp2.containsKey(leftChar) && mp1.get(leftChar) < mp2.get(leftChar))
                {
                    check--;
                }
                left++;
            }
            right++;
        }
        return (minLen == Integer.MAX_VALUE) ? "" : s.substring(start, start + minLen);
    }
}
